jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java

```diff
@@ -104,15 +128,285 @@ public Object getAttribute(String name)
+                return _map.remove(name);
+
+            return _map.put(name, attribute);
+        }
+
+        @Override
+        public Object getAttribute(String name)
+        {
+            return _map.get(name);
+        }
+
+        @Override
+        public Set<String> getAttributeNames()
+        {
+            return _names;
+        }
+
+        @Override
+        public void clearAttributes()
+        {
+            _map.clear();
+        }
+
+        public int size()
+        {
+            return _map.size();
+        }
+
+        public Set<Map.Entry<String, Object>> getAttributes()
+        {
+            return _map.entrySet();
+        }
+
+        @Override
+        public String toString()
+        {
+            return _map.toString();
+        }
+
+        public void addAll(Attributes attributes)
+        {
+            for (String name : attributes.getAttributeNames())
+                setAttribute(name, attributes.getAttribute(name));
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return _map.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (o instanceof Attributes)
+            {
+                Attributes a = (Attributes)o;
+                for (Map.Entry<String, Object> e : _map.entrySet())
+                {
+                    if (!Objects.equals(e.getValue(), a.getAttribute(e.getKey())))
+                        return false;
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by another Attributes instance, which is treated as immutable, but with a
+     * ConcurrentHashMap used as a mutable layer over it.
+     */
+    class Layer implements Attributes
+    {
+        private static final Object REMOVED = "REMOVED";
+        private final Attributes _persistent;
+        private final java.util.concurrent.ConcurrentMap<String, Object> _map = new ConcurrentHashMap<>();
+        private final Set<String> _names;
+
+        public Layer(Attributes persistent)
+        {
+            _persistent = persistent;
+            _names = new LayeredNamesSet();
+        }
+
+        @Override
+        public Object removeAttribute(String name)
+        {
+            if (_persistent.getAttributeNames().contains(name))
+            {
+                Object v = _map.put(name, REMOVED);
+                if (v == REMOVED)
+                    return null;
+                if (v != null)
+                    return v;
+                return _persistent.getAttribute(name);
+            }
+            Object v = _map.remove(name);
+            return v == REMOVED ? null : v;
+        }
+
+        @Override
+        public Object setAttribute(String name, Object attribute)
+        {
+            if (attribute == null)
+                return removeAttribute(name);
+            Object v = _map.put(name, attribute);
```

Comment:
I would have expected that if you set an unrelated attribute it would be persisted onto the nested `Attributes`. For example if you're doing a `INCLUDE` dispatch and then you set a custom attribute you would expect it to be available after the include.

 Or is this `Layer` mechanism not intended to be used for dispatched attributes?

Reply:
The use-case for this is the `ContextHandler` which contains a `Context`. Attributes set on the `ContextHandler` persist over multiple start/stop cycles, whilst attributes set on the `Context` are not and can be treated as a layer above the `ContextHandler` attributes.

It may be that `Layer` is only really specific to `ContextHandler` so I could move it there, but I wanted to take it out of `ContextHandler` (which currently has a half hearted layer impl) to reduce the volume of the `ContextHandler` class.